code/Chapter4/Exercise_solutions/a_param_test_normal_function.cpp

```diff
@@ -0,0 +1,38 @@
+#include <gtest/gtest.h>
+
+// Define your function to be tested here.
+int Sum(int a, int b) {
+    return a + b;
+}
+
+// Define a test fixture class
+class ParameterizedTest : public testing::TestWithParam<std::pair<int, int>> {
+};
+
+// Define the test case with the parameterized test
+TEST_P(ParameterizedTest, TestSum) {
+    // Get the parameter values
+    int a = GetParam().first;
+    int b = GetParam().second;
+
+    // Call your normal function
+    int result = Sum(a, b);
+
+    // Perform assertion
+    ASSERT_EQ(a + b, result);
+}
+
+// Define the test data
+INSTANTIATE_TEST_SUITE_P(Default, ParameterizedTest, testing::Values(
+    std::make_pair(1, 1),
+    std::make_pair(2, 3),
+    std::make_pair(-5, 10)
+));
```

Comment:
Just to confirm I've understood it right, if we had 4 different `TEST_P` cases, we will need the corresponding 4 different `INSTANTIATE_TEST_SUITE_P` to define the input values for them, right?

Reply:
@dalonsoa The answer is not straightforward Yes or No. IT would depend on case to case basis.

For example, if you have to functions that you want to test for multiple values (or say parameters for now), you can use the same class and same `INSTANTIATE_TEST_SUITE_P` will work. 

However, if you want to test a function with just different values, you might have to create another class and Test suite for this. This is because `INSTANTIATE_TEST_SUITE_P` instantiates a test suite which can have multiple different tests/functions.

Both concepts are used in the following code
```cpp
#include <gtest/gtest.h>

// Define your function to be tested here.
int Sum(int a, int b) {
    return a + b;
}


// function to multiply two numbers
double Multiply(double a, double b) {
    return a * b;
}

// Function to calculate a raised to the power of b (integer) without using pow function
double Power(double a, int b) {
    double result = 1;
    for(int i = 0; i < b; i++) {
        result *= a;
    }
    return result;
}

// Define a test fixture class
class ParameterizedTest : public testing::TestWithParam<std::pair<int, int>> {
};


// Define the test case with the parameterized test
TEST_P(ParameterizedTest, TestSum) {
    // Get the parameter values
    int a = GetParam().first;
    int b = GetParam().second;

    // Call your normal function
    int result = Sum(a, b);

    // Perform assertion
    ASSERT_EQ(a + b, result);
}

// Define the test data
INSTANTIATE_TEST_SUITE_P(Default, ParameterizedTest, testing::Values(
    std::make_pair(1, 1),
    std::make_pair(2, 3),
    std::make_pair(-5, 10)
));



// Define the test case with the parameterized test
TEST_P(ParameterizedTest, TestMultiply) {
    // Get the parameter values
    int a = GetParam().first;
    int b = GetParam().second;

    // Call your normal function
    int result = Multiply(a, b);

    // Perform assertion
    ASSERT_EQ(a * b, result);
}


// Define a test fixture class
class ParameterizedTest_Power : public testing::TestWithParam<std::tuple<double, int, double>> {
};


//Check if the power function works fine for different values of a and b
TEST_P(ParameterizedTest_Power, TestPowerFun)
{
    // Get the parameter values
    double a = std::get<0>(GetParam());
    int b = std::get<1>(GetParam());
    double answer = std::get<2>(GetParam());

    // Call your normal function
    double result = Power(a, b);

    // Perform assertion
    ASSERT_EQ(answer, result);
}

// Define the test data
INSTANTIATE_TEST_SUITE_P(PowTest, ParameterizedTest_Power, testing::Values(
    std::make_tuple(1, 1, 1),
    std::make_tuple(2, 3, 8),
    std::make_tuple(2.5, 2, 6.25)
));

int main(int argc, char** argv) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
```

Reply:
Very clear explanation, many thanks!